app/javascript/components/ReduxFormAutoComplete/index.js

```diff
@@ -0,0 +1,70 @@
+import React from 'react'
+import { Dropdown, Menu, Item, Autocomplete, Field as GardenField } from '@zendeskgarden/react-dropdowns'
+import debounce from 'lodash.debounce'
+
+const ReduxFormAutocomplete = ({ dataSource, input: { onChange }, searchField }) => {
+  const searchData = dataSource.map(data => data[searchField])
+  // Selected Item is the String
+  const [selectedItem, setSelectedItem] = React.useState('')
+  const [inputValue, setInputValue] = React.useState('')
+  const [matchingOptions, setMatchingOptions] = React.useState(searchData)
+
+  /**
+   * Debounce filtering
+   */
+  const filterMatchingOptionsRef = React.useRef(
```

Comment:
why are we using a `useRef` hook here? afaik the useRef hook here will create a mutable variable for the lifetime of this component. I don't see it being mutated anywhere and it is also a const.

Reply:
I think this was part of the React Garden examples, and the variable is mutated when `inputValue` is changed in the `useEffect` block. The `const` might be so it can't be re-assigned?

Reply:
For reference: https://garden.zendesk.com/react-components/dropdowns/#!/Autocomplete%20usage

Reply:
From my understanding `debounce` will return a function `A`, and `useRef` here will return a mutable ref object whose .current property is initialized to `A`. So in the `useEffect` block, we are essentially just executing function `A` right?

I just don't see the point of using `useRef` here, unless I am missing something. 

e.g. you could do the same thing by just removing it:

```
const filterMatchingOptionsRef = debounce(value => {
      const matchingOptions = searchData.filter(data => {
        return (
          data
            .trim()
            .toLowerCase()
            .indexOf(value.trim().toLowerCase()) !== -1
        )
      })

      setMatchingOptions(matchingOptions)
    }, 300)

React.useEffect(() => {
    filterMatchingOptionsRef(inputValue)
  }, [inputValue])
```

Reply:
actually @samzx , I did some research and it looks like the `useRef` is to prevent the debounce function from resetting between renders. Because if it did, there would be no debouncing.

a much more in-depth explanation here: https://overreacted.io/making-setinterval-declarative-with-react-hooks/